### ISSUE TITLE
V3.1.4 on z/OS abends S0C4 for function fetch_both and returns a python UTF decode error for function field name

### ISSUE THREAD

**Issue by pjfarleyiii** [ibmdb/python-ibmdb#852]
The errors described in this issue are occurring on the IBM Zxplore open learning system for z/OS (anyone can register to join at [Zxplore Learning System](https://ibmzxplore.influitive.com/)). I am not an admin of that system and have no access to any software on that system except what the IBM Zxplore admins provide.  I have no authority to install or upgrade software on that system, only the IBM Zxplore admins have that authority.

Please advise what I can communicate to the Zxplore admins to get assistance in resolving these issues.

- Operating System Name: z/OS V2.5
- db2level output from clidriver if in use:
- Target Db2 Server Version:  V12
- Python Version:  Python 3.9.2
- ibm_db version: 3.1.4
- For non-Windows, output of below commands:
  uname  OS/390
  uname -m  3931
- Value of below environment variables if set:
  IBM_DB_HOME:   DSNC10
  PATH:  /usr/lpp/IBM/cvg/go/bin:/usr/lpp/IBM/cyp/v3r9/pyz/bin:/usr/lpp/IBM/zoautil/bin:/usr/lpp/db2c10/jdbc/bin:/bin:/z/bin:.:/usr/lpp/java/J8.0_64/bin:/usr/lpp/zLiberty/V8R5/wlp/bin:/z/z02446/bin:/z/rocket/tools/bin:/usr/lpp/ported/bin:/usr/lpp/IBM/cnj/v18r0/IBM/node-latest-os390-s390x/bin:/usr/lpp/IBM/cnj/v18r0/njsc/bin:/global/node/bin:/usr/lpp/mqm/V9R1M1/web/bin:/usr/lpp/IBM/cobol/igyv6r4/bin
  LIB/LD_LIBRARY_PATH/DYLD_LIBRARY_PATH:  /usr/lpp/db2c10/jdbc/lib

- Test script to reproduce the problem.   
- For fetch_both() attached script db_fetch_both_err.py as text file
- For field_name() attached script db_fetch_name_err.py as text file

[db_fetch_both_err.py.txt](https://github.com/ibmdb/python-ibmdb/files/11537717/db_fetch_both_err.py.txt)

[db_fetch_name_err.py.txt](https://github.com/ibmdb/python-ibmdb/files/11537720/db_fetch_name_err.py.txt)

Steps to Reproduce:

1. python3 db_fetch_both_err.py
2. python3 db_fetch_name_err.py

Output from test scripts for functions fetch_both() and field_name() (Zxplore userid masked):

```
/z/zxxxxx > python3.9 db_fetch_both_err.py
Test start fetch_both
Connection done
exec_immediate done
CEE3204S The system detected a protection exception (System Completion Code=0C4).
         From compile unit TOROLABA:./Objects/object.c at entry point PyObject_Hash at statement 769 at compile unit offset +000000001C3A89B4 at entry offset +0000000000000074 at address 000000001C3A89B4.
Segmentation fault

/z/zxxxxx > python3.9 db_fetch_name_err.py
Test start field_name
Connection done
exec_immediate done
first fetch done
row 1 returned 3 columns: Traceback (most recent call last):
  File "/z/zxxxxx/db_fetch_name_err.py", line 41, in <module>
    colname = ibm_db.field_name(stmt, _i)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc5 in position 0: invalid continuation byte
```

**Comment by pjfarleyiii** [comment]
Has anyone been able to review these errors enough to know if they are environmental to the Zxplore system or actual ibmdb issues?  Would appreciate knowing if there is any further information I could provide to assist with analysis and remediation.

**Comment by Earammak** [comment]
@pjfarleyiii I tried on my local machine and i didn't face any problem. Looks like issue is because of encoding. Can you please check what is the encoding you have.   
Thanks

**Comment by pjfarleyiii** [comment]
I am not sure which encoding the DB2 subsystem is using, but the general encoding I use is "IBM1047" or in python "cp1047".

If you can tell me how to query it I will try to find out what the Zxplore DB2 subsystem is using by default, but I am kind of assuming it is probably code page 037.

**Comment by pjfarleyiii** [comment]
I found out how to display the CCSID's in use by DB2 at the Zxplore site.  SPUFI output from that site follows.

```
   SELECT GETVARIABLE('SYSIBM.SYSTEM_ASCII_CCSID')   ASCII_CCSID                                   
        , GETVARIABLE('SYSIBM.SYSTEM_EBCDIC_CCSID')  EBCDIC_CCSID                                  
        , GETVARIABLE('SYSIBM.SYSTEM_UNICODE_CCSID') UNICODE_CCSID                                 
    FROM SYSIBM.SYSDUMMY1                                                                          
   ;                                                                                               
---------+---------+---------+---------+---------+---------+---------+---------+---------+---------
ASCII_CCSID                                                                       EBCDIC_CCSID     
---------+---------+---------+---------+---------+---------+---------+---------+---------+---------
819,65534,65534                                                                   1047,65534,65534 
                  
------+---------+-
 UNICODE_CCSID    
------+---------+-
 367,1208,1200    
                  
DSNE610I NUMBER OF ROWS DISPLAYED IS 1                                                             
```

**Comment by Earammak** [comment]
@pjfarleyiii, I hope now are you able to run your test scripts correctly.
To check the existing file setting you can check from: 
  ls -aLT <file name>
Please refer this [page](https://www.ibm.com/docs/en/zos/2.5.0?topic=descriptions-chtag-change-file-tag-information) to know about setting of change file tag information.

Please let us know if you need any further help.

Thanks

**Comment by pjfarleyiii** [comment]
@Earammak I do know about file tagging and my scripts are tagged IBM1047.  They fail as coded.

```
/z/z99999 > ls -laT db*.py
t IBM-1047    T=on  -rw-r--r--   1 Z02446   IPGROUP     1428 Mar  3 18:44 db_fetch_both_err.py
t IBM-1047    T=on  -rw-r--r--   1 Z02446   IPGROUP     1918 May 22 18:11 db_fetch_name_err.py
```

**Comment by Earammak** [comment]
@pjfarleyiii, I'm able to run in my environment. I used the same python script as your shared. 
bash-4.3$ ls -laT fetch_both_err.py
t IBM-1047    T=on  -rw-r--r--   1 TS3596   PDUSER      1306 Jun 16 05:22 fetch_both_err.py
bash-4.3$ ls -laT fetch_name_error_852.py
t IBM-1047    T=on  -rw-r--r--   1 TS3596   PDUSER      1821 Jun 16 05:18 fetch_name_error_852.py
bash-4.3$ python3 --version
Python 3.9.5
bash-4.3$ uname -a
OS/390 RS25 28.00 04 3931
bash-4.3$ python3  fetch_both_err.py
Test start fetch_both
Connection done
exec_immediate done
first fetch done
Test end fetch_tuple
bash-4.3$ python3  fetch_name_error_852.py
Test start field_name
Connection done
exec_immediate done
first fetch done
Test end field_name
bash-4.3$
[fetch_both_err.py.txt](https://github.com/ibmdb/python-ibmdb/files/11768711/fetch_both_err.py.txt)
[fetch_name_error_852.py.txt](https://github.com/ibmdb/python-ibmdb/files/11768712/fetch_name_error_852.py.txt)

Please let me know if there is any discrepancy in my script.

Thanks

**Comment by pjfarleyiii** [comment]
@Earammak the script outputs you got are not correct, even though they did not get any error.  The "fetch_name" script is supposed to print the number of columns returned for each row and the names returned by the "fetch_name()" function for each column of each row.  Your execution printed nothing, as if your DB2 data table is empty.  Please try it with a DB2 table that contains some data and see what you get.

The same is true for the "fetch_both" script, if the DB2 table is not empty it should print the row number of each row returned by the "fetch_both()" function.  Again, please try that script with a DB2 table that has some data in it and see what you get.

**Comment by Earammak** [comment]
@pjfarleyiii, Find the below logs for the above-mentioned scripts.
[issue_852_logs.zip](https://github.com/ibmdb/python-ibmdb/files/11814156/issue_852_logs.zip)

I'm able to run them successfully, and don't see any problem.
Can you please share here the issue you are facing here for your latest scripts.

Thanks

**Comment by pjfarleyiii** [comment]
See my original report at the top.  The "fetch_both" script abends at the first call to the "fetch_both()" function with S0C4 (addressing exception) in something called "TOROLABA" somewhere deep inside of the python runtime code.  The "fetch_name" script ends with a decoding error at the first call to the "fetch_name()" function.  I'll repeat my sample output here:

```
/z/zxxxxx > python3.9 db_fetch_both_err.py
Test start fetch_both
Connection done
exec_immediate done
CEE3204S The system detected a protection exception (System Completion Code=0C4).
         From compile unit TOROLABA:./Objects/object.c at entry point PyObject_Hash at statement 769 at compile unit offset +000000001C3A89B4 at entry offset +0000000000000074 at address 000000001C3A89B4.
Segmentation fault

/z/zxxxxx > python3.9 db_fetch_name_err.py
Test start field_name
Connection done
exec_immediate done
first fetch done
row 1 returned 3 columns: Traceback (most recent call last):
  File "/z/zxxxxx/db_fetch_name_err.py", line 41, in <module>
    colname = ibm_db.field_name(stmt, _i)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc5 in position 0: invalid continuation byte
```

**Comment by Earammak** [comment]
@pjfarleyiii Since you are observing this problem only on IBM Zxplore z/OS system and it is working on your local system. 
  Please check with IBM Zxplore z/OS system administrator on this issue that how we can resolve it. 

Thanks

**Comment by pjfarleyiii** [comment]
@Earammak  You have it all wrong.  I have NO LOCAL SYSTEM to run these scripts on I ONLY HAVE ZXPLORE!!!

WHAT CAN I DO TO GET MORE INFORMATION FOR YOU TO BE ABLE TO SOLVE THE ISSUE I SEE ON ZXPLORE?

The Zxplore admins have said they will cooperate to help me solve the problem, but I have to know what to ask them to do or what I have to do to give you the information you need to solve the problem.  I sent you a dump, is that not enough information?  What more do you need to solve the issue?

Please help me solve this issue.

**Comment by jthyssenrocket** [comment]
@pjfarleyiii 

Was a SVC dump taken? If yes, can we get that, please? If no, can you set a SLIP for the S0C4 and capture a dump for us, please?

**Comment by pjfarleyiii** [comment]
I cannot do that (I do not have the privileges needed), but I can ask the Zxplore sysadmin to set a SLIP and capture the dump for you. I can create the S0C4 on demand, so it should not take too long to get it.  I will post it in a zip file when I have received it from the Zxplore team.

**Comment by jthyssenrocket** [comment]
Thanks! Do you know if the Zxplore team have plans to upgrade Python, as v3.9.2 is  ~2 years old?

**Comment by pjfarleyiii** [comment]
As I understand it from the Zxplore admins, it isn't entirely up to them - they use some "common" base system generated by some other IBM group (possibly WAZI?) and have to wait for that "common" base to be updated, they aren't always allowed to do upgrades on their own.

Meanwhile I am still waiting on a response to my request to set up a SLIP trap.  As this python package isn't on their "we will positively support this" list, it may take a bit longer to get the dump.  They did say they would try to help, but we have to depend on their availability to coordinate with me to get the dump, and there are time zone issues.  I think my main contact is on Zulu time (UTC 0 offset) and I am UTC -4 (EDT).

**Comment by jthyssenrocket** [comment]
They are likely using the IBM ADCD distribution also used by other IBM  solutions. 
The latest release of ADCD is from May 2023, is built on RSU2303 and contains IBM Python 3.11.0.

If you can find HLQ ADCD in ISPF 3.4 you can determine the ADCD build from the second LQ. For example, the existence of ADCD.Z25D suggests it is the May 2023 distribution mentioned above. 

Can you also confirm you're running on real hardware? I see machine type 3931 from uname, so likely a z16? The ADCD distribution is often used by emulated environments which have machine types 1090-1092. 

The reason why I am asking is that I had another customer report a similar issue a while ago. They opened a support case with IBM Open Enterprise Python support, but the issue was never diagnosed as a refresh of their system to latest maintenance resolved the issue.

**Comment by pjfarleyiii** [comment]
Yes, I believe they are running on a real z16, there was a hardware upgrade last month or the month before where they said they were going to a z16.  No datasets under HLQ ADCD though.

**Comment by pjfarleyiii** [comment]
Progress report: The Zxplore admins set up a SLIP as requested and I generated the S0C4, but the SLIP did not catch it yet.

I will continue to work with the Zxplore admins to get the dump generated from the SLIP and post here again when we have it.

**Comment by pjfarleyiii** [comment]
Progress report: Unfortunately no more progress as yet.  Still trying to get the repeatable S0C4 to trigger the SLIP.  Administrators have been busy with other Zxplore issues, but I hope to have something available soon.

**Comment by denisgaebler** [comment]
Maybe its a similar issue, but any ideas why running on z/OS (printing ibm_db.conn_errormsg() ) would create a conversion error:
```
Traceback (most recent call last):
  File "/u/myid/python/testdb2.py", line 20, in <module>
    print( ibm_db.conn_errormsg() )
           ^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 0: ordinal not in range(128)
```

And given the error reported above also occurs at conversion, is there a chance that something returned by the odbc driver is EBCDIC and as such not converted correctly? 0xe4 is 'ä' but I recall from using 1047 codepage at 3270 terminals that could be a bracket such as [] or {} displayed as 'ä'.

It could be environmental, wouldn't be the first time that anything else than cp037 would not work. Any environment variable or odbc setting I can look at or try to get rid of the conversion error?

It would be awkward, if one had to do ibm_db.conn_errormsg().decode(???) to get rid of the error.

**Comment by pjfarleyiii** [comment]
@denisgaebler I had that same problem when first trying to issue an "exec_immediate()" to DB2.  Below is what I had to code to see the error message from the failed operation.  Part of the error message seems to be encoded in UTF-8, part seems to be encoded in IBM-1047.  Perhaps something similar will work for a connect error for you.

```
    try:
        stmt = ibm_db.exec_immediate(conn,sql_stmtz)
    except:
        print("exec_immediate failed:")
        errmsg = str(ibm_db.stmt_errormsg().encode("ISO8859-1").decode("IBM1047"))
        errmsg = str(errmsg[:-13]) + str(errmsg[-13:].encode("IBM1047").decode("utf-8"))
        print(errmsg)
    else:
        print("exec_immediate done")
```

**Comment by daveyc** [comment]
@pjfarleyiii previously opened a [ticket](https://github.com/daveyc/pyzfile/issues/2) for pyzfile related to encoding issues on Z Xplore. It was bizarre as the encoding was performed in the Python runtime. I wasn't able to recreate the problem on local systems. This leads me to form the opinion with a high degree of probability that Python 3.9.2 on Z Xplore is broken. At the very least Python on Z Xplore should be upgraded to v3.9.16 PH53440 for there to be any point in investigating this issue. If we get a dump all that will do is confirm that the ABEND0C4 is caused by bad encoding.

**Comment by pjfarleyiii** [comment]
@daveyc, As I responded to you on IBM-MAIN, IMHO it is not a useless exercise to nail down the root cause of these errors.  If that root cause turns out to be a broken python version on the Zxplore system, that provides us with proof to goad the Zxplore admins to make a request that their version of python be upgraded.  As far as I am aware, they do not actually completely control the contents and versions of products on the system they are using to reach out to those who wish to learn more about the z/OS environment and quirks.  Their mandate is to teach and help people learn.   My quest here is to make a useful product (ibmdb) work in that teaching environment for my own learning benefit and the benefit to other students seeking to learn there.

**Comment by daveyc** [comment]
@pjfarleyiii I think we should close this issue. Firstly, and most importantly, Python 3.9.2 is no longer available so we can't test it a Rocket. I'm also dubious about the Python distribution on the Z XPlore system. The traceback contains path names such as `TOROLABA` (Toronto Laboratories) that would suggest this may be beta code. 

```
    6     PyObject_Hash                                                                                                             
                      TOROLABA:./Objects/object.c                         /VERSYSB/usr/lpp/IBM/cyp/v3r9/pyz/lib/libpython3.9.so     
    7     PyDict_SetItem                                                                                                          
```

> My quest here is to make a useful product (ibmdb) work in that teaching environment for my own learning benefit and the benefit to other students seeking to learn there.

100%. But there is nothing the maintainers of this repo can do on an unsupported and broken Python.

**Comment by Earammak** [comment]
@pjfarleyiii Can you please provide update on this?

**Comment by bimalkjha** [comment]
@daveyc Another user hit the similar problem. Detail is as below:
```
Message without dump:
CEE3204S The system detected a protection exception (System Completion Code=0C4)
.
         From compile unit TOROLABA:./Objects/typeobject.c at entry point PyType_Ready@AF337_43 at 
statement 6365 at compile unit offset +00000000232109EE at entry offset +000000000000001E at address 00000000232109EE.
[1] + Done(139) python odbc.fetcheboth.py
  50662733      Segmentation violation  /usr/bin/python  

Requested system dump to be attached.

The python appears to fail on ibm_db.fetch_both(stmt)
although when I replace it with ibm_db.fetch_tuple(stmt) the program works.

Environment
z/OS 2.5
DB2 z 12 Function Level 500
Python 3.10.9
ibm_db 3.1.4

Python script is encoded with cp1047 and terminal set to IBM-1047.
```
They have shared the SVCDUMP files that you can access from ftp2.rocketsoftware.com by searching [TS013773510](https://ftp2.rocketsoftware.com/human.aspx?r=425867524&arg06=979578606&arg12=filelist). Please download the dump and have a look into it. Thanks.

**Comment by pjfarleyiii** [comment]
@Earammak , my status is that I am still awaiting a response from the Zxplore admins about whether the SLIP and the SVCDUMP DASD space are available and active.  I have also tried generating the dump using 

```export _CEE_RUNOPTS='FILETAG(AUTOCVT,AUTOTAG) TRAP(OFF,NOSPIE) TER(UAIMM)'```

but I have yet to see an SVCDUMP get created from the abend process.  I am afraid this issue is just not very high on the list of things the Zxplore admins are concerned with.

If the other dump that was submitted is not sufficient to resolve the issue I can continue to try to get an SVCDUMP from Zxplore.  Please keep me posted on whether or not I should continue to try to get this dump.

**Comment by pjfarleyiii** [comment]
@bimalkjha , thank you for letting us know that another user has encountered the same error and was able to provide a dump for debugging.

**Comment by jthyssenrocket** [comment]
@pjfarleyiii 

What value do you have for CURRENTAPPENSCH in your ODBC INI file (pointed to by environment variable DSNAOINI)?

Can you try with CURRENTAPPENSCH=UNICODE (or ASCII) ? With this setting the encode/decode code is not needed.

**Comment by daveyc** [comment]
I diagnosed the SVC dump and it was related to character encoding. I was able to recreate the ABEND on Rocket systems with several versions of Python. As @jthyssenrocket has suggested, this is a user configuration problem. Using a correct ODBC configuration which specifies `CURRENTAPPENSCH=UNICODE` (or ASCII) fixes the problem. 

This is documented in the README with links but perhaps it would be clearer if it were included as part of a `z/OS Configuration` section. Unfortunately, there doesn't seem to be a way of retrieving the value of `CURRENTAPPENSCH` programmatically, otherwise we could check the pre-conditions when the package is initialized.  

> z/OS: In case of any issues using Python on z/OS, refer to this [file](https://github.com/ibmdb/python-ibmdb/blob/master/install.md) and also refer to this [doc](https://github.com/ibmdb/node-ibm_db#configure-odbc-driver-on-zos).

FWIW, for future reference setting SLIP dumps is overkill. Set the following environment variable to enable dynamic LE dumps (SYSMDUMP).

`export _CEE_RUNOPTS='FILETAG(AUTOCVT,AUTOTAG) TER(UADUMP) DYNDUMP(*USERID.CEEDUMP,DYNAMIC,TDUMP)'`

**Comment by daveyc** [comment]
The underlying issue originates from the fact that PyUnicode_FromString returns NULL when it can't generate a Unicode object. Subsequently, this NULL pointer is passed to PyDict_SetItem, leading to a segmentation fault. Although this problem arises from an incorrect configuration, it might be beneficial to invest some effort in integrating "panic" assertions that test conditions categorized as "should not happen." This precautionary measure could help avert occurrences such as 0C4 abends, which consume valuable time during the processes of uploading and troubleshooting.

https://github.com/ibmdb/python-ibmdb/blob/be7d1cc0836f16a36866fadd50e1d6b595885cbb/ibm_db.c#L9546

**Comment by bimalkjha** [comment]
@pjfarleyiii Could you please apply set CURRENTAPPENSCH as suggested in above post and verify?
@daveyc Thanks for identifying the root cause! It would be good if you can open a PR with desired "panic" assertions.
Also, do you think changing line no 9548 from `if (value == NULL) {` to `if (value == NULL || key == NULL) {` will help here? Thanks.

**Comment by daveyc** [comment]
@bimalkjha I think we should investigate if there is any way on z/OS that we can query the configuration and fail fast if it's not configured for `UNICODE` or `ASCII`. I don't know how to do that but we should reach out to our DB2 experts. @jthyssenrocket do you know anybody?

**Comment by jthyssenrocket** [comment]
I'll reach out to my Db2 ODBC z/OS contact at SVL

**Comment by pjfarleyiii** [comment]
> @pjfarleyiii
> 
> What value do you have for CURRENTAPPENSCH in your ODBC INI file (pointed to by environment variable DSNAOINI)?
> 
> Can you try with CURRENTAPPENSCH=UNICODE (or ASCII) ? With this setting the encode/decode code is not needed.

The Zxplore odbc.ini already uses CURRENTAPPENSCH=ASCII, paste below.

~~~
/z/z99999 > echo $DSNAOINI
/etc/odbc.ini
/z/z99999 > cat $DSNAOINI    
; This is a comment line...
; Example COMMON stanza
[COMMON]
MVSDEFAULTSSID=DBCG
CONNECTTYPE=1
MULTICONTEXT=2
CURRENTAPPENSCH=ASCII
FLOAT=IEEE
[DBCG]
AUTOCOMMIT=1
MVSATTACHTYPE=CAF
PLANNAME=DSNACLI
~~~

**Comment by jthyssenrocket** [comment]
```
Please also post the result of:

set |grep BPX
set |grep CEE
set |grep STEPLIB
set |grep TAG
set |grep DSNAOINI
```

**Comment by pjfarleyiii** [comment]
@jthyssenrocket , here are the requested "set | echo" results, working on the rerun with diagnostics next.

```
/z/z99999 > set |grep BPX
_BPXK_AUTOCVT=ON
/z/z99999 > set |grep CEE
STEPLIB=DSNC10.DBCG.SDSNEXIT:DSNC10.SDSNLOAD:DSNC10.SDSNLOD2:CEE.SCEERUN:IGY640.SIGYCOMP
_CEE_RUNOPTS='FILETAG(AUTOCVT,AUTOTAG) TERMTHDACT(UADUMP) ABTERMENC(ABEND)'
_LD_SYSLIB='//'\''CEE.SCEELKEX'\''://'\''CEE.SCEELKED'\''://'\''SYS1.CSSLIB'\''://'\''CBC.SCCNOBJ'\'''
/z/z99999 > set |grep STEPLIB
STEPLIB=DSNC10.DBCG.SDSNEXIT:DSNC10.SDSNLOAD:DSNC10.SDSNLOD2:CEE.SCEERUN:IGY640.SIGYCOMP
/z/z99999 > set |grep TAG
_CEE_RUNOPTS='FILETAG(AUTOCVT,AUTOTAG) TERMTHDACT(UADUMP) ABTERMENC(ABEND)'
_TAG_REDIR_ERR=txt
_TAG_REDIR_IN=txt
_TAG_REDIR_OUT=txt
/z/z99999 > set |grep DSNAOINI
DSNAOINI=/etc/odbc.ini
```

**Comment by jthyssenrocket** [comment]
Sorry, my mistake: DSNAO64T is in SDSNLOD2 (not SDSNLOAD). However, we need the DD:DSNAOTRC file first. 

There is an alternative set of instructions here:
https://www.ibm.com/docs/en/db2-for-zos/12?topic=odt-capturing-odbc-diagnostic-trace-information-in-zos-unix-environment
but it does not work in my environment. 

If it doesn't work, I'll need some help from IBM Db2 ODBC dev to get the ODBC diag trace captured, as it contains information about the options used including application encoding scheme.

**Comment by jthyssenrocket** [comment]
The alternative instructions actually works for me from TSO OMVS, but not from a putty ssh session.

**Comment by pjfarleyiii** [comment]
I will try the alternative instructions from a USS login and reply back.  I am assuming that DSNAO64T is what I should use for this test, correct?

**Comment by jthyssenrocket** [comment]
Sorry, if it wasn't clear: there are two alternative ways to collect the trace

Option 1: add the DIAG* keywords and optionally allocate DD DSNAOTRC file. It should be written during processing. 
I connect to USS using ssh, and when running the python program it just write a file named DD:DSNAOTRC (I don't explicitly allocate it). 
```
export DSNAOINI=/path/to/odbc_with_diagtrace.ini
python ...
./dsnao64t fmt DD\:DSNAOTRC dsnaotrc.txt   # format the trace 

```
Option 2: remove the DIAG* keywords, and use dsnao64t:
```
# odbc.ini without DIAG* keywords
export DSNAOINI=/path/to/odbc_without_diagtrace.ini
./dsnao64t on
python ... 
./dsnao64t dmp dmp.trc               # this should write a dmp.trc file
./dsnao64t fmt dmp.trc dmp.txt    # this should format the trace into a text file 
```

The error message "Not enough memory to hold the trace." suggests that your TSO region size (or imposed OMVS limits) is too small.

**Comment by pjfarleyiii** [comment]
OK, I tried option 2 from a USS ssh login, results pasted below.  I will try from an OMVS shell in TSO but I suspect the REGION for TSO is smaller than the USS limit on this "student" system.

```
/z/z99999 > ./dsnao64t on
Trace is turned on
/z/z99999 > python3 db_fetch_both_err-2.py
Test start fetch_both
Connection done
exec_immediate done
CEE3204S The system detected a protection exception (System Completion Code=0C4).
         From compile unit TOROLABA:./Objects/object.c at entry point PyObject_Hash at statement 769 at compile unit offset +000000001C4A79B4 at entry offset +0000000000000074 at address 000000001C4A79B4.
Segmentation fault
/z/z99999 > ./dsnao64t dmp odbc-dump.trc
Could not close odbc-dump.trc.
/z/z99999 > ./dsnao64t fmt odbc-dump.trc odbc-dump.txt
Not enough memory to hold the trace.
```

**Comment by pjfarleyiii** [comment]
Alternate SQL SELECT results pasted below.  You can ignore the final traceback error since variable "rowno" was never set due to the exec_immediate failure.

```
/z/z99999 > python3 db_fetch_both_err-2.py
Test start fetch_both
Connection done
exec_immediate failed:
{DB2 FOR OS/390}{ODBC DRIVER}{DSN12015}
 DSNT408I SQLCODE = -199, ERROR:  ILLEGAL USE OF KEYWORD SCHEME.  TOKEN ENCODING
          COMPATIBILITY WAS EXPECTED
 DSNT418I SQLSTATE   = 42601 SQLSTATE RETURN CODE
 DSNT415I SQLERRP    = DSNHPARS SQL PROCEDURE DETECTING ERROR
 DSNT416I SQLERRD    = 2  0  0  -1  28  506 SQL DIAGNOSTIC INFORMATION
 DSNT416I SQLERRD    = X'00000002'  X'00000000'  X'00000000'  X'FFFFFFFF'
          X'0000001C'  X'000001FA' SQL DIAGNOSTIC INFORMATION
  ERRLOC=5:10:2 SQLCODE=-199
Traceback (most recent call last):
  File "/z/z99999/db_fetch_both_err-2.py", line 45, in <module>
    print("Test end fetch_both, {} rows printed".format(rowno))
NameError: name 'rowno' is not defined
```

**Comment by jthyssenrocket** [comment]
`1 ('1047    ',)`

1047 is EBCDIC (US codepage), so the python program is still using EBCDIC despite setting CURRENTAPPENSCH=UNICODE in the ODBC ini file, which is what I suspected. This is why the ODBC diag trace is critical, so we can see why ODBC decides to use EBCDIC instead of the specified UNICODE. 

Would it be possible to retry "option 2" with your new larger TSO REGION size? 

FWIW, On  my system the same program returns:

`#1 {'': '1208    ', 0: '1208    '}`

Codepage 1208 is unicode, so this is the expected result.

**Comment by pjfarleyiii** [comment]
I will do that when I get back home later today.

**Comment by jthyssenrocket** [comment]
@pjfarleyiii 

What is the tagging of your ODBC ini file?

ls -lTE $DSNAOINI

**Comment by pjfarleyiii** [comment]
@jthyssenrocket , Encoding of odbc.ini file follows.

```
/z/z99999 > ls -lTE $DSNAOINI
t IBM-1047    T=on  -rw-r--r--  --s-  1 OMVSKERN OMVSGRP      198 Jul  4 07:18 /etc/odbc.ini
```

**Comment by jthyssenrocket** [comment]
Annoyingly, I've had the note about the ODBC ini file in my Python ibmdb on z/OS presentation for several years, but this is the first time I've seen the issue materialize as an S0C4. 

Unfortunately it appears there is no easy way to programmatically check the application encoding scheme after connecting to Db2, but before issuing the first SQL statement.

**Comment by daveyc** [comment]
@pjfarleyiii can you please prod the ZXplore admins to upgrade Python? A browse through to Python SDK fixlist flags multiple fixes for encoding issues. You may find that other committers are not so accommodating to help debug issues like this.

**Comment by pjfarleyiii** [comment]
@daveyc, I have already asked about upgrading the python installation on Zxplore.  The answer I received is that the Zxplore z/OS image is derived from another IBM team's "base" image, and the version of python which is available on Zxplore is installed by that other team in their image.  When that other team's "base" image is updated and merged into the Zxplore system is when the version of python on Zxplore will be updated.

**Comment by daveyc** [comment]
> In fact, the existence of the ibmdb package and your pyzfile package on Zxplore are due to the admin there accommodating my request to install them at the global level when he had some time available.

I understand your perspective. It's unfortunate that your workplace doesn't provide a more advanced USS development environment, considering you're challenging the notion that experienced individuals can't pick up new skills.

**Comment by jthyssenrocket** [comment]
@pjfarleyiii 

I don't have an /etc/odbc.ini file on my system, so it must be something the Zxplore (or the owners of the installation they inherited) has created. I can't find any sample odbc.ini file in /samples or in /usr/lpp/db2 either. 

Would it be worthwhile to ship a sample ODBC ini file with ibm_db that gets installed into site_packages with the proper tagging? 

There is an Aha idea open to have ibm_db included in the IBM python distribution (see https://ibm-z-software-portal.ideas.ibm.com/ideas/PY-I-7). This would also be an opportunity to ship a sample ODBC ini file with proper tagging.

**Comment by pjfarleyiii** [comment]
@daveyc, yes, I am bummed that my employer doesn't even have python installed, much less a USS development environment.  Some things we just have to "work around" on our own to keep on learning.

Continuously learning new things is what keeps old-timers like me alive and kickin'.  I've always said that if I enjoy what I do 50% of the time, I am better off than 80% of working stiffs, and the other 50% is just what you have to put up with in any employment position.  Learning new stuff is well up there in the 50% I enjoy.

**Comment by jthyssenrocket** [comment]
@pjfarleyiii 

Not sure, but I think you should be able to reproduce the table above if you start "sh". I rarely use IBM /bin/sh as it is too painful :)

I am not able to find odbc.ini on my system (neither in /etc, /samples or /usr/lpp/db2)

**Comment by pjfarleyiii** [comment]
@jthyssenrocket, 1000% agreed, IBM /bin/sh is WAY too painful.  If I get some time I will try to see if ```_BPXK_AUTOCVT=ALL``` makes any difference, just out of curiosity.

**Comment by pjfarleyiii** [comment]
Status report from Zxplore: Their ```/etc/odbc.ini``` file has been changed to the "mixed" tag and all my scripts are operating correctly without using my private copy of that fle, so at least no one else at Zxplore will run into the same issue that I did.

Again, my thanks to everyone here for your patience and help.
